Get full response from PulseAudio server

breiti opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
For larger responses from the PulseAudio server there's a chance that
`recv()` is called before the data has been received, leaving us with a
truncated result as the remaining text to be received is silently
ignored.

This change makes sure that all data received until the configured tcp
timeout elapses is evaluated, at the cost of making all requests that
expect a result slower.

Note that the result text contains information about the number of
modules loaded, which should make it possible to determine when the
returned text ends without waiting for the timeout. Until this is fixed,
this changes makes the behavior more reliable.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

breiti committed:
Get full response from PulseAudio server

For larger responses from the PulseAudio server there's a chance that
recv() is called before the data has been received, leaving us with a
truncated result as the remaining text to be received is silently
ignored.

This change makes sure that all data received until the configured tcp
timeout elapses is evaluated, at the cost of making all requests that
expect a result slower.

Note that the result text contains information about the number of
modules loaded, which should make it possible to determine when the
returned text ends without waiting for the timeout. Until this is fixed,
this change makes the behavior more reliable.

homeassistant commented:
Hi @breiti, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#33360).

Once you do that we will be able to review and accept this pull request.

Thanks!

breiti committed:
Stop waiting for response from PulseAudio server

The response from PulseAudio server for the 'list modules' command
states how many modules are loaded right at the beginning.  We can use
this information to detect when all modules haven been received and
prevent the need to wait until the tcp timeout occurs.

MartinHjelmare commented on homeassistant/components/pulseaudio_loopback/switch.py:
```diff
@@ -84,47 +86,74 @@ def __init__(self, host, port, buff_sz, tcp_timeout):
+        while remaining_timeout > 0:
+            ready = select.select([sock], [], [], remaining_timeout)
+            if ready[0]:
+                rcv_buffer = sock.recv(self._buffer_size)
```
Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices

MartinHjelmare reviewed (changes requested):
Thanks for the PR. Unfortunately we can't accept any (big) changes to this integration until the device specific interface code has been extracted to a 3rd party library.

breiti commented:
Extracting library in https://github.com/home-assistant/core/pull/34965, making this PR obsolete.
